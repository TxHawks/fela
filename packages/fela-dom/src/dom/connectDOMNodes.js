/* @flow */
import objectEach from 'fast-loops/lib/objectEach'
import arrayEach from 'fast-loops/lib/arrayEach'

import {
  clusterCache,
  cssifySupportRules,
  sheetMap,
  RULE_TYPE,
} from 'fela-utils'

import initDOMNode from './initDOMNode'
import findDOMNodes from './findDOMNodes'

import type { DOMRenderer } from '../../../../flowtypes/DOMRenderer'

export default function connectDOMNodes(renderer: DOMRenderer): void {
  renderer.nodes = findDOMNodes()

  const cacheCluster = clusterCache(
    renderer.cache,
    renderer.mediaQueryOrder,
    renderer.supportQueryOrder
  )

  const baseNode = renderer.nodes[RULE_TYPE]

  objectEach(sheetMap, (type, key) => {
    if (cacheCluster[key].length > 0) {
      initDOMNode(renderer.nodes, baseNode, cacheCluster[key], type)
    }
  })

  const support = cssifySupportRules(cacheCluster.supportRules)

  if (support) {
    initDOMNode(renderer.nodes, baseNode, support, RULE_TYPE, '', true)
  }

  const mediaKeys = Object.keys({
    ...cacheCluster.supportMediaRules,
    ...cacheCluster.mediaRules,
  })

  arrayEach(mediaKeys, media => {
    if (
      cacheCluster.mediaRules[media] &&
      cacheCluster.mediaRules[media].length > 0
    ) {
      initDOMNode(
        renderer.nodes,
        baseNode,
        cacheCluster.mediaRules[media],
        RULE_TYPE,
        media
      )
    }

    if (cacheCluster.supportMediaRules[media]) {
      const mediaSupport = cssifySupportRules(
        cacheCluster.supportMediaRules[media]
      )

      if (mediaSupport.length > 0) {
        initDOMNode(
          renderer.nodes,
          baseNode,
          mediaSupport,
          RULE_TYPE,
          media,
          true
        )
      }
    }
  })
}
