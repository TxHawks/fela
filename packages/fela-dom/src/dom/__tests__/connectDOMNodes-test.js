import { html as beautify } from 'js-beautify'
import { RULE_TYPE } from 'fela-utils'

import connectDOMNodes from '../connectDOMNodes'

describe('Connecting DOM nodes', () => {
  it('should initialize nodes', () => {
    const node = document.createElement('style')
    node.setAttribute('data-fela-type', 'RULE')
    node.type = 'text/css'

    document.head.appendChild(node)

    const renderer = {
      cache: {
        colorred: {
          className: 'a',
          selector: '.a',
          declaration: 'color:red',
          media: '',
          support: '',
          type: RULE_TYPE,
        },
        colorblue: {
          className: 'b',
          selector: '.b',
          declaration: 'color:blue',
          media: '(min-height: 800px)',
          support: '',
          type: RULE_TYPE,
        },
        coloryellow: {
          className: 'c',
          selector: '.c',
          declaration: 'color:yellow',
          media: '',
          support: '(display: flex)',
          type: RULE_TYPE,
        },
        colorgreen: {
          className: 'd',
          selector: '.d',
          declaration: 'color:green',
          media: '(min-height: 800px)',
          support: '(display: flex)',
          type: RULE_TYPE,
        },
      },
    }

    connectDOMNodes(renderer)

    expect(renderer.nodes.RULE).toEqual(node)
    expect([
      Object.keys(renderer.nodes),
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      }),
    ]).toMatchSnapshot()
  })
})
