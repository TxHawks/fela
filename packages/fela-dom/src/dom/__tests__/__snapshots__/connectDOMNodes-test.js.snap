// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connecting DOM nodes should initialize nodes 1`] = `
Array [
  Array [
    "RULE",
    "RULEsupport",
    "RULE(min-height: 800px)",
    "RULE(min-height: 800px)support",
  ],
  "<html>

<head>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\">
    .a {
      color: red
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" data-fela-support=\\"true\\">
    @supports (display: flex) {
      .c {
        color: yellow
      }
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" media=\\"(min-height: 800px)\\">
    .b {
      color: blue
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\" data-fela-support=\\"true\\" media=\\"(min-height: 800px)\\">
    @supports (display: flex) {
      .d {
        color: green
      }
    }
  </style>
</head>

<body></body>

</html>",
]
`;
